pkg/util/util.go

```diff
@@ -33,3 +36,65 @@ func GetRequestHost(req *http.Request) string {
 	}
 	return host
 }
+
+// RefreshValue contains a composite cookie refresh value setting
+type RefreshValue struct {
+	RefreshDuration *time.Duration
+	RefreshPercent  uint8
+}
+
+// Set a value of a refresh
+func (r *RefreshValue) Set(s string) error {
+	origS := s
+	s = strings.TrimSpace(s)
+	if strings.HasSuffix(s, "%") {
+		i, err := strconv.ParseInt(s[:len(s)-1], 10, 8)
+		if err != nil {
+			return fmt.Errorf("percentage \"%s\" must be a valid integer: %s", origS, err)
+		}
+		if i < 1 || i > 100 {
+			return fmt.Errorf("percentage \"%s\" must be an integer between 1 and 100", origS)
+		}
+		r.RefreshPercent = uint8(i)
+	} else {
+		dur, err := time.ParseDuration(s)
+		if err != nil {
+			return fmt.Errorf("\"%s\" must be a valid time duration: %s", origS, err)
+		}
+		if dur < time.Duration(0) {
+			return fmt.Errorf("\"%s\" must be a non-negative time duration", origS)
+		}
+		r.RefreshDuration = &dur
+	}
+	return nil
+}
+
+// IsDurationBased returns true if duration
+func (r *RefreshValue) IsDurationBased() bool {
+	return r.RefreshDuration != nil
+}
+
+// IsDisabled returns true if refresh is disabled
+func (r *RefreshValue) IsDisabled() bool {
+	return r.RefreshDuration != nil && *r.RefreshDuration == time.Duration(0)
+}
+
+// ShouldRefresh tells if a session cookie should refresh
+func (r *RefreshValue) ShouldRefresh(now time.Time, created time.Time, expires time.Time) bool {
```

Comment:
NIT: we should rename `now` maybe

We are checking if we should refresh at the given point in time.  It just happens in must use cases we pass in `time.Now()`

Reply:
Open to thoughts on this idea... on the fence on it:

Should this be a method of a session, which takes in a `*RefreshValue` --

so: `func (s *SessionState) ShouldRefresh(rv *util.RefreshValue)`

I think it puts the ownership of the logic in the right place -- the session is then asking "ShouldRefresh with this RefreshValue?"

Reply:
Actually, I like this, cause then the `SessionState` in its implementation of this can handle the `session.RefreshToken == nil` case and return `false` in the method rather than us needing to manage it and have a more complex conditional logic above in the call.

Reply:
Why rename `now`? It is named exactly for what it represents - the point in time that is considered "right now" and corresponds to the `time.Now()`.

I'll rename `RefreshValue.ShouldRefresh` to `RefreshValue.HasExpired` and add `SessionState.ShouldRefresh` to wrap with additional logic. While the fields of the `RefreshValue` are public, the interpretation of expiration of is controlled by the option itself, based on **its** internal state.

Reply:
If this is always `now`, do we need it as an argument? Or is this just dependency injection pattern of the current time to be used later in test cases once added?

Something like this might result in a cleaner method interface while giving the freedom to stub time logic in the test suite without needing to include a whole time mocking library: https://stackoverflow.com/a/41661462

Reply:
It may be useful to check expiration against some "other" now or have "now" shared between several methods, which will need to work on the same time instant. Hardcoding "time.Now()" in a bunch of methods will produce inconsistent results (e.g. `now` from one method != `now` from another method)